src/Symfony/Component/PasswordHasher/Hasher/NativePasswordHasher.php

```diff
@@ -73,10 +73,14 @@ public function __construct(int $opsLimit = null, int $memLimit = null, int $cos
 
     public function hash(string $plainPassword): string
     {
-        if ($this->isPasswordTooLong($plainPassword) || ((string) \PASSWORD_BCRYPT === $this->algorithm && 72 < \strlen($plainPassword))) {
+        if ($this->isPasswordTooLong($plainPassword)) {
             throw new InvalidPasswordException();
         }
 
+        if ((string) \PASSWORD_BCRYPT === $this->algorithm && (72 < \strlen($plainPassword) || false !== strpos($plainPassword, "\0"))) {
+            $plainPassword = base64_encode(hash('sha512', $plainPassword, true));
```

Comment:
This creates a string that is longer than 72

```
$ php -r "echo strlen(base64_encode(hash('sha512', 'test', true))), PHP_EOL;"
88
```
```suggestion
            $plainPassword = hash('sha256', $plainPassword);
```

Reply:
This still creates a string longer that 72.
Yet we don't care about the truncation, because bcrypt will truncate on its own.
What we care about is providing maximum entropy to bcrypt while never providing it with a nul byte.
base64_encode achieves that.

Reply:
> This still creates a string longer that 72.

Wrong :) But still, a truncated sha512 is stronger than a sha256.

Reply:
Thanks for clarifying that the truncation is intentional